gpu/test/base/gpu_executor.cu

```diff
@@ -142,6 +146,37 @@ TEST_F(GpuExecutor, CopiesDataFromGpu)
     gpu->free(orig);
 }
 
+/* Properly checks if it works only when multiple GPUs exist */
+TEST_F(GpuExecutor, PreservesStreamSettings)
+{
+    auto last_device = gko::GpuExecutor::get_num_devices() - 1;
+    cudaSetDevice(last_device);
+    auto orig = gpu->alloc<int>(2);
+    int current_stream;
+    cudaGetDevice(&current_stream);
+    ASSERT_EQ(current_stream, last_device);
+
+    gpu->free(orig);
+    cudaGetDevice(&current_stream);
+    ASSERT_EQ(current_stream, last_device);
+}
+
+TEST_F(GpuExecutor, CopiesDataFromGpuToGpu)
+{
+    int copy[2];
+    auto orig = gpu->alloc<int>(2);
+    cudaSetDevice(0);
+    init_data<<<1, 1>>>(orig);
+
+    auto copy_gpu2 = gpu2->alloc<int>(2);
+    gpu2->copy_from(gpu.get(), 2, orig, copy_gpu2);
+    cpu->copy_from(gpu2.get(), 2, copy_gpu2, copy);
+
+    EXPECT_EQ(5, copy[0]);
+    ASSERT_EQ(2, copy[1]);
```

Comment:
Maybe we can make sure that the data is actually on `gpu2` by launching a kernel on it, similar to [`check_data`](https://github.com/ginkgo-project/ginkgo/blob/f2e5cb2af8912d61174852da839cec3bab9d04c6/gpu/test/base/gpu_executor.cu#L110).

Reply:
Isn't this what this does, in a way similar to `CopiesDataFromGpu`, because it does an explicit copy from `gpu2` to `cpu`.

Reply:
I cannot verify this as I don't have a system with 2 GPUs, but I have a feeling that if you apply this [patch](https://github.com/ginkgo-project/ginkgo/files/1811242/wrong.patch.txt) (which obviously makes the code wrong), all tests (except RunsOnProperDevice) will still pass.

Which means that the tests are not actually testing what you want them to test.

Reply:
After testing it looks like you are right. This is actually fairly weird because the data is allocated to the wrong GPU for `gpu2`, in which case the pointer passed for copying is completely broken but CUDA doesn't seem to care so much about that?

Reply:
Actually, it's not weird at all, and the pointers are not broken - if you apply this patch you never change GPUs. `gpu` and `gpu2` are actually referencing the same GPU (the one with id 0, since we never use `setDevice`), and gpu -> gpu2 copy is copying data between two memory locations in the same GPU.

Reply:
A `cudaMemcpyPeer` happens between `gpu` and `gpu2` which have different IDs though. And it copies to a pointer allocated on `gpu`.

Reply:
Ah yes, you're right. I guess the reason for this weird behavior is because the system you're using has uniform memory addressing, so CUDA runtime can deduce the location of the pointers just by looking at their values, and ignores device ids completely.

Reply:
Actually this is fairly annoying, `check_data` also does not work because it uses the pointer and if the data is allocation on `GPU 0`, it launches the kernel there... Even using `cudaSetDevice` does not seem to force execution there. I will try to do something else.

Reply:
> if the data is allocated on GPU 0, it launches the kernel there

Are you sure about this? I don't think CUDA runtime is allowed to override your decision about where to launch the kernel. Can you verify this with `nvprof` or `nvvp`?

Which system are you running this on? Could you paste the output of `nvidia-smi`?

Reply:
I explained that wrongly my bad. I was actually correcting my post. I meant that if, as with the patch, the device id doesn't change anyway everything is launched and allocated on `GPU 0` and it runs fine. If I do somehow force execution of the check kernel on `GPU 1`, it anyway just fetches the data and also runs fine... I also did not see a way to disable UMA, nor did I see a way to check from the `check_data` kernel itself the device used to ensure it's actually the second GPU. I think I will also add the kernel call that checks we really are launching stuff on the correct GPU, like the previous test. I cannot see anything else.

Reply:
> If I do somehow force execution of the check_data kernel on GPU 1, it anyway just fetches the data and also runs fine...

That is what I figured is going on. Still, knowing more about the system you're running it on (at least having the output of `nvidia-smi`) would help a lot in explaining what's going on.

Btw, is it by any chance a K80 system?

Reply:
This is actually 4 K40s and I also tried another with two P100s, with similar results. I have access to another one with older tesla I could try there.

The systems:
>+-----------------------------------------------------------------------------+
| NVIDIA-SMI 384.66                 Driver Version: 384.66                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla K40m          Off  | 00000000:03:00.0 Off |                    0 |
| N/A   31C    P0    63W / 235W |      0MiB / 11439MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla K40m          Off  | 00000000:04:00.0 Off |                    0 |
| N/A   29C    P0    61W / 235W |      0MiB / 11439MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla K40m          Off  | 00000000:82:00.0 Off |                    0 |
| N/A   30C    P0    62W / 235W |      0MiB / 11439MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  Tesla K40m          Off  | 00000000:83:00.0 Off |                    0 |
| N/A   28C    P0    64W / 235W |      0MiB / 11439MiB |     81%      Default |
+-------------------------------+----------------------+----------------------+

>+-----------------------------------------------------------------------------+
| NVIDIA-SMI 384.66                 Driver Version: 384.66                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla P100-PCIE...  Off  | 00000000:04:00.0 Off |                    0 |
| N/A   33C    P0    29W / 250W |      0MiB / 16276MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla P100-PCIE...  Off  | 00000000:82:00.0 Off |                    0 |
| N/A   33C    P0    26W / 250W |      0MiB / 16276MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

Reply:
In that case, I'm as surprised as you are...

I though maybe it was a K80 board, because that one has 2 GPUs which share the same memory - that would explain the results.

Reply:
Yes it is, it seems to behave as if direct access of peer memory is set as if `cudaDeviceEnablePeerAccess` was used, except it was telling me it is not. See [doc](http://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__UNIFIED.html#group__CUDART__UNIFIED)

Reply:
Well, they never explicitly state that you won't be able to access peer memory if you don't use `cudaDeviceEnablePeerAccess`. Maybe they now support it by default and that function is implemented as no-op.

Reply:
Yeah, but the problem is the `cudaDeviceDisablePeerAccess` was giving an error stating it's not enabled. Anyway, as I said I added a check to ensure we're not cheating, much like for the test above that one.